File: internal/eval/jsonpath.go
package eval

import (
	"encoding/json"
	"fmt"
	"strconv"
	"strings"
)

// segment is one step of a parsed JSONPath: a map key or an array index.
type segment struct {
	key   string
	index int
	isKey bool
}

// parsePath parses the JSONPath subset the DSL supports: a leading "$", dot
// keys ($.a.b), bracket keys ($['a.b']), and array indices ($.a[0]). Filters,
// wildcards, and recursive descent are intentionally unsupported.
func parsePath(path string) ([]segment, error) {
	if path == "" || path[0] != '$' {
		return nil, fmt.Errorf("path %q must start with $", path)
	}
	var segs []segment
	i := 1
	for i < len(path) {
		switch path[i] {
		case '.':
			i++
			if i < len(path) && path[i] == '.' {
				return nil, fmt.Errorf("recursive descent (..) is not supported: %q", path)
			}
			start := i
			for i < len(path) && isIdentByte(path[i]) {
				i++
			}
			if i == start {
				return nil, fmt.Errorf("empty key after '.' in %q", path)
			}
			segs = append(segs, segment{key: path[start:i], isKey: true})
		case '[':
			seg, next, err := parseBracket(path, i)
			if err != nil {
				return nil, err
			}
			segs = append(segs, seg)
			i = next
		default:
			return nil, fmt.Errorf("unexpected %q at position %d in %q", path[i], i, path)
		}
	}
	return segs, nil
}

func parseBracket(path string, i int) (segment, int, error) {
	i++ // past '['
	if i < len(path) && (path[i] == '\'' || path[i] == '"') {
		quote := path[i]
		i++
		start := i
		for i < len(path) && path[i] != quote {
			i++
		}
		if i >= len(path) {
			return segment{}, 0, fmt.Errorf("unterminated quoted key in %q", path)
		}
		key := path[start:i]
		i++ // past closing quote
		if i >= len(path) || path[i] != ']' {
			return segment{}, 0, fmt.Errorf("expected ] after quoted key in %q", path)
		}
		return segment{key: key, isKey: true}, i + 1, nil
	}
	start := i
	for i < len(path) && path[i] != ']' {
		i++
	}
	if i >= len(path) {
		return segment{}, 0, fmt.Errorf("unterminated [ in %q", path)
	}
	n, err := strconv.Atoi(strings.TrimSpace(path[start:i]))
	if err != nil {
		return segment{}, 0, fmt.Errorf("array index %q in %q is not an integer", path[start:i], path)
	}
	return segment{index: n}, i + 1, nil
}

func isIdentByte(b byte) bool {
	return b == '_' || b == '-' ||
		(b >= 'a' && b <= 'z') || (b >= 'A' && b <= 'Z') || (b >= '0' && b <= '9')
}

// queryJSON walks path into a JSON body. The bool is false when the body is
// valid JSON but the path names nothing; a non-JSON body is an error.
func queryJSON(body []byte, path string) (any, bool, error) {
	segs, err := parsePath(path)
	if err != nil {
		return nil, false, err
	}
	if len(body) == 0 {
		return nil, false, nil
	}
	var root any
	if err := json.Unmarshal(body, &root); err != nil {
		return nil, false, fmt.Errorf("response body is not JSON: %w", err)
	}
	cur := root
	for _, s := range segs {
		if s.isKey {
			m, ok := cur.(map[string]any)
			if !ok {
				return nil, false, nil
			}
			if cur, ok = m[s.key]; !ok {
				return nil, false, nil
			}
			continue
		}
		arr, ok := cur.([]any)
		if !ok {
			return nil, false, nil
		}
		idx := s.index
		if idx < 0 {
			idx += len(arr)
		}
		if idx < 0 || idx >= len(arr) {
			return nil, false, nil
		}
		cur = arr[idx]
	}
	return cur, true, nil
}

File: sdk-python/src/flowbench/jsonpath.py
"""The JSONPath subset internal/eval/jsonpath.go supports: a leading "$", dot
keys ($.a.b), bracket keys ($['a.b']), and array indices ($.a[0], $.a[-1]).
Filters, wildcards, and recursive descent are intentionally unsupported --
this is a deliberate port of Go's restricted grammar, not a general JSONPath
library, so Python accepts exactly what the Go engine accepts, never more.
"""

import json

from .errors import FlowExecutionError


class _Key:
  __slots__ = ("key",)

  def __init__(self, key):
    self.key = key


class _Index:
  __slots__ = ("index",)

  def __init__(self, index):
    self.index = index


def _is_ident_byte(c):
  return c == "_" or c == "-" or c.isalnum()


def parse_path(path):
  if path == "" or path[0] != "$":
    raise FlowExecutionError(f"path {path!r} must start with $")

  segments = []
  i = 1
  n = len(path)
  while i < n:
    c = path[i]
    if c == ".":
      i += 1
      if i < n and path[i] == ".":
        raise FlowExecutionError(f"recursive descent (..) is not supported: {path!r}")
      start = i
      while i < n and _is_ident_byte(path[i]):
        i += 1
      if i == start:
        raise FlowExecutionError(f"empty key after '.' in {path!r}")
      segments.append(_Key(path[start:i]))
    elif c == "[":
      seg, i = _parse_bracket(path, i)
      segments.append(seg)
    else:
      raise FlowExecutionError(f"unexpected {c!r} at position {i} in {path!r}")
  return segments


def _parse_bracket(path, i):
  i += 1  # past '['
  n = len(path)
  if i < n and path[i] in ("'", '"'):
    quote = path[i]
    i += 1
    start = i
    while i < n and path[i] != quote:
      i += 1
    if i >= n:
      raise FlowExecutionError(f"unterminated quoted key in {path!r}")
    key = path[start:i]
    i += 1  # past closing quote
    if i >= n or path[i] != "]":
      raise FlowExecutionError(f"expected ] after quoted key in {path!r}")
    return _Key(key), i + 1

  start = i
  while i < n and path[i] != "]":
    i += 1
  if i >= n:
    raise FlowExecutionError(f"unterminated [ in {path!r}")
  raw = path[start:i].strip()
  try:
    index = int(raw)
  except ValueError:
    raise FlowExecutionError(
      f"array index {raw!r} in {path!r} is not an integer"
    ) from None
  return _Index(index), i + 1


def query_json(body, path):
  """Returns (value, found). Raises FlowExecutionError on invalid JSON or a
  malformed path. found is False when the body is valid JSON but the path
  names nothing -- mirrors internal/eval/jsonpath.go's queryJSON exactly.
  """
  segments = parse_path(path)
  if not body:
    return None, False

  try:
    cur = json.loads(body)
  except (json.JSONDecodeError, UnicodeDecodeError) as e:
    raise FlowExecutionError(f"response body is not JSON: {e}") from e

  for seg in segments:
    if isinstance(seg, _Key):
      if not isinstance(cur, dict) or seg.key not in cur:
        return None, False
      cur = cur[seg.key]
    else:
      if not isinstance(cur, list):
        return None, False
      idx = seg.index
      if idx < 0:
        idx += len(cur)
      if idx < 0 or idx >= len(cur):
        return None, False
      cur = cur[idx]
  return cur, True
